optimism/src/mips/witness.rs

```diff
@@ -225,9 +264,94 @@ impl<Fp: Field> Env<Fp> {
         }
     }
 
-    pub fn step(&mut self) {
+    pub fn step(&mut self, config: VmConfiguration, start: &Start) {
         println!("instruction: {:?}", self.decode_instruction());
+
+        self.pp_info(config.info_at, start);
+
+        // Force stops at given iteration
+        if self.should_trigger_at(config.stop_at) {
+            self.halt = true;
+            return;
+        }
+
         // TODO
         self.halt = true;
     }
+
+    fn should_trigger_at(&self, at: StepFrequency) -> bool {
+        let m: u64 = self.instruction_counter as u64;
+        match at {
+            StepFrequency::Never => false,
+            StepFrequency::Always => true,
+            StepFrequency::Exactly(n) => n == m,
+            StepFrequency::Every(n) => m % n == 0,
+        }
+    }
+
+    // Compute memory usage
+    fn memory_usage(&self) -> String {
+        let total = self.memory.len() * PAGE_SIZE;
+        memory_size(total)
+    }
+
+    fn page_address(&self) -> (u32, usize) {
+        let address = self.instruction_pointer;
+        let page = (address >> PAGE_ADDRESS_SIZE) as u32;
+        let page_address = (address & (PAGE_ADDRESS_MASK as u32)) as usize;
+        (page, page_address)
+    }
+
+    fn get_opcode(&mut self) -> Option<u32> {
+        let (page_id, page_address) = self.page_address();
+        for (page_index, memory) in self.memory.iter_mut() {
```

Comment:
Nit: we're not mutating, so this can just be `iter`

Reply:
Done!